reactor-core/src/test/java/reactor/core/publisher/FluxWindowPredicateTest.java

```diff
@@ -42,6 +42,40 @@
 public class FluxWindowPredicateTest extends
                                      FluxOperatorTest<String, Flux<String>> {
 
+	@Test
+	public void windowWhileNoEmptyWindows() {
+		Flux.just("ALPHA", "#", "BETA", "#")
+		    .windowWhile(s -> !"#".equals(s))
+		    .flatMap(Flux::collectList)
+		    .as(StepVerifier::create)
+		    .assertNext(w -> assertThat(w).containsExactly("ALPHA"))
+		    .assertNext(w -> assertThat(w).containsExactly("BETA"))
+		    .verifyComplete();
+	}
+
+	@Test
+	public void windowUntilNoEmptyWindows() {
+		Flux.just("ALPHA", "#", "BETA", "#")
+		    .windowUntil("#"::equals)
+		    .flatMap(Flux::collectList)
+		    .as(StepVerifier::create)
+		    .assertNext(w -> assertThat(w).containsExactly("ALPHA", "#"))
+		    .assertNext(w -> assertThat(w).containsExactly("BETA", "#"))
+		    .verifyComplete();
+	}
```

Comment:
You might also want a couple more tests for intentionally empty windows:

```java
@Test
 public void windowWhileIntentionallyEmptyWindows() {
   Flux.just("ALPHA", "#", "BETA", "#", "#")
       .windowWhile(s -> !"#".equals(s))
       .flatMap(Flux::collectList)
       .as(StepVerifier::create)
       .assertNext(w -> assertThat(w).containsExactly("ALPHA"))
       .assertNext(w -> assertThat(w).containsExactly("BETA"))
       .assertNext(w -> assertThat(w).isEmpty())
       .verifyComplete();
 }

 @Test
 public void windowUntilIntentionallyEmptyWindows() {
   Flux.just("ALPHA", "#", "BETA", "#", "#")
       .windowUntil("#"::equals)
       .flatMap(Flux::collectList)
       .as(StepVerifier::create)
       .assertNext(w -> assertThat(w).containsExactly("ALPHA", "#"))
       .assertNext(w -> assertThat(w).containsExactly("BETA", "#"))
       .assertNext(w -> assertThat(w).isEmpty())
       .verifyComplete();
 }
```

Reply:
"intentionally empty windows": ok, scratch that PR, back to the drawing board

Reply:
so @nebhale this is absolutely 100% about not emitting an empty window on source completion? Is the following correct:

 - [ ] starting `windowUntil` with `cutBefore true` with a delimiter will produce an empty initial window, the next one will have the delimiter + whatever values
 - [ ] starting `windowWhile` with a non-matching value will produce an empty window
 - [ ] putting several delimiters in a row in a `windowWhile` will produce intermediate empty windows?
 - [ ] only case where you don't want empty windows is `windowUntil` ending with a delimiter and `windowWhile` ending with a delimiter

Reply:
It's a good question, and I don't think I've ever thought about it in this way (you should see my face contorting as I think this over...).

I view the boundary predicate as indicating the end of a window, but _not implicitly the start of the next window_.  Only something that is explicitly window-related (i.e. new element or another boundary) can start a new window (and possibly close it immediately).

To be more concrete, I'd say that for `windowWhile()` the boundary predicate **can** implicitly open a _previous_ window, but **cannot** implicitly open a _next_ window.  For `windowUntil()` the rule holds (it can open _previous_ but not _next_) however when the **boundary is processed as an item** (either cut before or cut after) it may **also** force a new window (either _previous_ or _next_) to be opened.

Addressing your checklist, all four are `true`, which I think (hope) is consistent with what I've stated above.  In other words, I believe that the only contentious bit of the implementation in `master` is that boundaries implicitly open the _next_ window.

```java
@Test
public void windowUntilCutBefore() {
    Flux.just("#", "#", "ALPHA", "#", "#", "BETA", "#", "#")
        .windowUntil("#"::equals, true)
        .flatMap(Flux::collectList)
        .as(StepVerifier::create)
        .assertNext(w -> assertThat(w).isEmpty())
        .assertNext(w -> assertThat(w).containsExactly("#"))
        .assertNext(w -> assertThat(w).containsExactly("#", "ALPHA"))
        .assertNext(w -> assertThat(w).containsExactly("#"))
        .assertNext(w -> assertThat(w).containsExactly("#", "BETA"))
        .assertNext(w -> assertThat(w).containsExactly("#"))
        .assertNext(w -> assertThat(w).containsExactly("#"))
        .verifyComplete();
}

@Test
public void windowUntilCutAfter() {
    Flux.just("#", "#", "ALPHA", "#", "#", "BETA", "#", "#")
        .windowUntil("#"::equals, false)
        .flatMap(Flux::collectList)
        .as(StepVerifier::create)
        .assertNext(w -> assertThat(w).containsExactly("#"))
        .assertNext(w -> assertThat(w).containsExactly("#"))
        .assertNext(w -> assertThat(w).containsExactly("ALPHA", "#"))
        .assertNext(w -> assertThat(w).containsExactly("#"))
        .assertNext(w -> assertThat(w).containsExactly("BETA", "#"))
        .assertNext(w -> assertThat(w).containsExactly("#"))
        .verifyComplete();
}

@Test
public void windowWhile() {
    Flux.just("#", "#", "ALPHA", "#", "#", "BETA", "#", "#")
        .windowWhile(s -> !"#".equals(s))
        .flatMap(Flux::collectList)
        .as(StepVerifier::create)
        .assertNext(w -> assertThat(w).isEmpty())
        .assertNext(w -> assertThat(w).isEmpty())
        .assertNext(w -> assertThat(w).containsExactly("ALPHA"))
        .assertNext(w -> assertThat(w).isEmpty())
        .assertNext(w -> assertThat(w).containsExactly("BETA"))
        .assertNext(w -> assertThat(w).isEmpty())
        .verifyComplete();
}
```